python/redmonster/physics/zfinder.py

```diff
@@ -4,42 +4,109 @@
 import matplotlib as m
 from matplotlib import pyplot as p
 
+import multiprocessing
+import sys
+import time
+
 from redmonster.datamgr.ssp_prep import SSPPrep
 from redmonster.physics.misc import poly_array, two_pad
 from redmonster.datamgr.io2 import read_ndArch, write_chi2arr
 
 # Assumes all templates live in $REDMONSTER_DIR/templates/
 
+def _zchi2(arg) :
+    return zchi2_single_template(**arg)
+
+def _zchi2_no_poly(arg) :
+    return zchi2_single_template_no_poly(**arg)
+
+
+def zchi2_single_template(j,poly_fft, t_fft, t2_fft, data_fft, ivar_fft, pmat_pol, bvec_pol, chi2_0, chi2_null, num_z, npixstep, zminpix, flag_val_neg_model) :
+    npoly=poly_fft.shape[0] # degree+1 of polynomial
+    fftnaxis1=t_fft.shape[0] # number of redshift bins
+    pmat = pmat_pol.copy() # precomputed block for polynomial terms , n.zeros( (npoly+1, npoly+1, fftnaxis1), dtype=float)
+    bvec = bvec_pol.copy() # precomputed vector block for polynomial terms , n.zeros( (npoly+1, fftnaxis1), dtype=float)
+
+    # fill matrix
+    pmat[0,0] = n.fft.ifft(t2_fft * ivar_fft.conj()).real
+    bvec[0]   = n.fft.ifft(t_fft * data_fft.conj()).real
+    for ipos in range(npoly):
+        pmat[ipos+1,0] = pmat[0,ipos+1] = n.fft.ifft(t_fft*poly_fft[ipos].conj()).real
+
+    # solve for each z
+
+    zchi2arr=n.zeros((num_z))
+    zwarning=n.zeros((num_z))
+    for l in n.arange(num_z)*npixstep:
+        try : # try to solve for this redshift
+            f = linalg.solve(pmat[:,:,l+zminpix],bvec[:,l+zminpix])
+
+            zchi2arr[(l//npixstep)] = chi2_0 - n.dot(n.dot(f,pmat[:,:,l+zminpix]),f) # is this true ?????
+            if f[0]<0 :
+                zwarning[(l//npixstep)] = int(zwarning[(l//npixstep)]) | flag_val_neg_model
+                zchi2arr[(l//npixstep)] = chi2_null
+                try:
+                    n.dot(n.dot(f,pmat[:,:,l+zminpix]),f)
+                except Exception as e:
+                    print("Except: %r" % e)
+                    zchi2arr[(l//npixstep)] = chi2_null
+        except : # failure
+            #print "failure"
+            #print sys.exc_info()
+            zchi2arr[(l//npixstep)] = chi2_null
+    return j,zchi2arr,zwarning
+
+def zchi2_single_template_no_poly(j,t_fft, t2_fft, data_fft, ivar_fft, chi2_0, num_z, npixstep, zminpix, flag_val_neg_model) :
+
+    a = n.fft.ifft(t2_fft * ivar_fft.conj()).real
+    b = n.fft.ifft(t_fft * data_fft.conj()).real
+
+    ii = zminpix+n.arange(num_z)*npixstep
+    f = (a[ii]!=0)*b[ii]/(a[ii]+(a[ii]==0))
+    zchi2arr = chi2_0 - a[ii]*f**2 # is this true ?????
+
+    zwarning=n.zeros((num_z))
+    zchi2arr[f<0] = chi2_0
+    zwarning[f<0] = flag_val_neg_model
+    return j,zchi2arr,zwarning
+
+
+
 class ZFinder:
-    
-    def __init__(self, fname=None, group=[0], npoly=None, zmin=None, zmax=None):
+    def __init__(self, fname=None, group=[0], npoly=None, zmin=None, zmax=None, nproc=1):
         self.fname = fname
-        if type(group) == 'list':
+        if type(group) == list:
```

Comment:
It's not - I need to do that.  Template groups were originally implemented as integers, as each template could be assigned to a single group.  I didn't feel like this was a complete solution to the problem, though, since the objects that are being classified aren't discrete, but exist on a continuum of sorts.  For example, maybe you want galaxy templates and quasar templates to flag against each other, but you don't want emission line galaxies to flag against galaxy or quasar.  There wasn't a way to have this happen with single groups, but assigned galaxy to group 1, ELG to groups [1,2], and quasar to group 2.  In order to do so, I sort of standardized that all groups are lists, and code that assumes that.